### PR TITLE
[CS] Invalidate VarDecls in targets that fail to type-check

### PR DESCRIPTION
**Pull request by hamishknight**
This avoids a crash that could occur when attempting to query their interface type later, which could cause us to attempt to type-check the Decl separately from its enclosing closure.

Eventually we also ought to use this to fill in ErrorTypes for expressions (since type-checking ought to only produce typed AST), but I'm leaving that as future work for now.

rdar://120012473

### CODE REVIEW

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

**Comment by hamishknight** [comment]
@swift-ci please SourceKit stress test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

**Comment by hamishknight** [comment]
@swift-ci please SourceKit stress test

File: lib/Sema/SyntacticElementTarget.cpp [hunk 1]
```diff
@@ -372,7 +396,58 @@ SyntacticElementTarget::walk(ASTWalker &walker) const {
     break;
   }
   case Kind::forEachStmt: {
-    if (auto *newStmt = getAsForEachStmt()->walk(walker)) {
+    // We currently do not type-check a for loop's BraceStmt as part of the
+    // SyntacticElementTarget, so we need to skip it here.
+    // TODO: We ought to be able to fold BraceStmt checking into the constraint
+    // system eventually.
```
Comment (xedin) [inline lib/Sema/SyntacticElementTarget.cpp:402]: Should this also respect `ignoreWhereClause` flag?

File: lib/Sema/TypeCheckConstraints.cpp [hunk 2]
```diff
@@ -571,6 +579,13 @@ Type TypeChecker::typeCheckParameterDefault(Expr *&defaultValue,
     DiagnosticTransaction diagnostics(ctx.Diags);
 
     TypeCheckExprOptions options;
+
+    // Avoid invalidating the AST since we'll fall through and try to type-check
+    // with an archetype contextual type opened. Note we also don't need to call
+    // `markInvalid` for the target below since we'll replace the expression
+    // with an ErrorExpr on failure anyway.
+    options |= TypeCheckExprFlags::AvoidInvalidatingAST;
+
```
[the remark this answers is not in the document]
Comment (xedin) [reply lib/Sema/TypeCheckConstraints.cpp:588]: I tried to do that initially when I worked on the feature, we effectively need a disjunction for contextual type here where the second element would be open when applied...

**Review by xedin — APPROVED** [review]

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
Updated to respect `ignoreWhereClause`, will rename `forEachStmt` in a follow-up

**Comment by xedin** [comment]
Thank you!

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by glessard** [comment]
@swift-ci please test macOS platform

**Comment by glessard** [comment]
@swift-ci please test macOS platform

**Pull request merged**